File: packages/shared/rulepack/rulepack.go
// Package rulepack loads and evaluates rp-* YAML rule packs (SPEC §1.4) and
// provides a client for the core rules-engine API (POST /v1/evaluate) with a
// local-embedded fallback so services run dev-standalone.
package rulepack

import (
	"bytes"
	"embed"
	"encoding/json"
	"fmt"
	"io/fs"
	"net/http"
	"os"
	"path/filepath"
	"strings"
	"time"

	"gopkg.in/yaml.v3"
)

//go:embed packs
var embedded embed.FS

// Pack is one rp-* rule-pack document (SPEC §1.4).
type Pack struct {
	ID                 string         `yaml:"id" json:"id"`
	Version            string         `yaml:"version" json:"version"`
	EffectiveFrom      string         `yaml:"effective_from" json:"effective_from"`
	EffectiveTo        *string        `yaml:"effective_to" json:"effective_to"`
	Status             string         `yaml:"status" json:"status"`
	SubjectToRegazette bool           `yaml:"subject_to_regazette" json:"subject_to_regazette"`
	Provenance         map[string]any `yaml:"provenance" json:"provenance"`
	Signed             map[string]any `yaml:"signed" json:"signed"`
	Rules              []Rule         `yaml:"rules" json:"rules"`
}

// Rule: when-clauses (context predicates) → then-effects (decision payload).
type Rule struct {
	ID   string         `yaml:"id" json:"id"`
	When map[string]any `yaml:"when" json:"when"`
	Then map[string]any `yaml:"then" json:"then"`
}

// Ref identifies a pack version, e.g. "rp-wht-2024@1.0.0".
func (p *Pack) Ref() string { return p.ID + "@" + p.Version }

// Parse decodes one pack YAML document.
func Parse(data []byte) (*Pack, error) {
	var p Pack
	if err := yaml.Unmarshal(data, &p); err != nil {
		return nil, fmt.Errorf("parse pack: %w", err)
	}
	if p.ID == "" || p.Version == "" {
		return nil, fmt.Errorf("pack missing id/version")
	}
	return &p, nil
}

// Load reads packs/<id>/<version>.yaml from dir; version "" = latest file.
func Load(dir, id, version string) (*Pack, error) {
	if version == "" {
		matches, err := filepath.Glob(filepath.Join(dir, "packs", id, "*.yaml"))
		if err != nil || len(matches) == 0 {
			return nil, fmt.Errorf("no versions of %s in %s", id, dir)
		}
		version = strings.TrimSuffix(filepath.Base(matches[len(matches)-1]), ".yaml")
	}
	data, err := os.ReadFile(filepath.Join(dir, "packs", id, version+".yaml"))
	if err != nil {
		return nil, err
	}
	return Parse(data)
}

// LoadEmbedded reads a pack from the embedded fallback copies.
func LoadEmbedded(id, version string) (*Pack, error) {
	if version == "" {
		entries, err := fs.ReadDir(embedded, "packs/"+id)
		if err != nil || len(entries) == 0 {
			return nil, fmt.Errorf("no embedded pack %s", id)
		}
		version = strings.TrimSuffix(entries[len(entries)-1].Name(), ".yaml")
	}
	data, err := embedded.ReadFile("packs/" + id + "/" + version + ".yaml")
	if err != nil {
		return nil, err
	}
	return Parse(data)
}

// TraceEntry records how one rule fired (or why not).
type TraceEntry struct {
	RuleID  string `json:"rule_id"`
	Matched bool   `json:"matched"`
	Reason  string `json:"reason,omitempty"`
}

// Decision is the merged outcome of evaluating a pack against a context.
type Decision struct {
	Pack  string         `json:"pack"`
	Attrs map[string]any `json:"decision"`
	Trace []TraceEntry   `json:"trace"`
}

// toFloat normalises YAML/JSON numerics.
func toFloat(v any) (float64, bool) {
	switch n := v.(type) {
	case int:
		return float64(n), true
	case int64:
		return float64(n), true
	case float64:
		return n, true
	case float32:
		return float64(n), true
	}
	return 0, false
}

// matchCond evaluates one when-clause key. Operator suffixes: __in, __ne,
// __lt, __lte, __gt, __gte, __exists.
func matchCond(key string, want any, ctx map[string]any) (bool, string) {
	base, op := key, "eq"
	for _, o := range []string{"__in", "__ne", "__lte", "__lt", "__gte", "__gt", "__exists"} {
		if strings.HasSuffix(key, o) {
			base, op = strings.TrimSuffix(key, o), strings.TrimPrefix(o, "__")
			break
		}
	}
	got, ok := ctx[base]
	switch op {
	case "exists":
		wantB, _ := want.(bool)
		return (ok == wantB) || want == nil, fmt.Sprintf("%s exists=%v", base, ok)
	case "in":
		list, _ := want.([]any)
		for _, item := range list {
			if fmt.Sprint(item) == fmt.Sprint(got) {
				return true, ""
			}
		}
		return false, fmt.Sprintf("%s=%v not in %v", base, got, want)
	case "eq":
		if fmt.Sprint(got) == fmt.Sprint(want) {
			return true, ""
		}
		if gf, gok := toFloat(got); gok {
			if wf, wok := toFloat(want); wok && gf == wf {
				return true, ""
			}
		}
		return false, fmt.Sprintf("%s=%v != %v", base, got, want)
	case "ne":
		m, _ := matchCond(base, want, ctx)
		return !m, fmt.Sprintf("%s ne %v", base, want)
	default: // numeric comparisons
		gf, gok := toFloat(got)
		wf, wok := toFloat(want)
		if !gok || !wok {
			return false, fmt.Sprintf("%s not numeric", base)
		}
		var m bool
		switch op {
		case "lt":
			m = gf < wf
		case "lte":
			m = gf <= wf
		case "gt":
			m = gf > wf
		case "gte":
			m = gf >= wf
		}
		if !m {
			return false, fmt.Sprintf("%s=%v fails %s %v", base, gf, op, wf)
		}
		return true, ""
	}
}

// Match reports whether all when-clauses hold for the context.
func (r *Rule) Match(ctx map[string]any) (bool, string) {
	for k, want := range r.When {
		ok, why := matchCond(k, want, ctx)
		if !ok {
			return false, why
		}
	}
	return true, ""
}

// Evaluate runs every rule and merges the `then` payloads of matching rules.
func Evaluate(p *Pack, ctx map[string]any) *Decision {
	d := &Decision{Pack: p.Ref(), Attrs: map[string]any{}}
	for i := range p.Rules {
		r := &p.Rules[i]
		ok, why := r.Match(ctx)
		d.Trace = append(d.Trace, TraceEntry{RuleID: r.ID, Matched: ok, Reason: why})
		if ok {
			for k, v := range r.Then {
				d.Attrs[k] = v
			}
		}
	}
	return d
}

// EngineClient calls the core rules-engine (SPEC §2) when reachable.
type EngineClient struct {
	BaseURL string // e.g. http://localhost:8020; empty disables
	Client  *http.Client
}

func (c *EngineClient) http() *http.Client {
	if c.Client != nil {
		return c.Client
	}
	return &http.Client{Timeout: 3 * time.Second}
}

// Evaluate POSTs /v1/evaluate {pack_id, version, context}.
func (c *EngineClient) Evaluate(packID, version string, ctx map[string]any) (*Decision, error) {
	if c.BaseURL == "" {
		return nil, fmt.Errorf("rules-engine base url empty")
	}
	body, _ := json.Marshal(map[string]any{"pack_id": packID, "version": version, "context": ctx})
	req, err := http.NewRequest(http.MethodPost, strings.TrimRight(c.BaseURL, "/")+"/v1/evaluate", bytes.NewReader(body))
	if err != nil {
		return nil, err
	}
	req.Header.Set("Content-Type", "application/json")
	resp, err := c.http().Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("rules-engine status %d", resp.StatusCode)
	}
	var out struct {
		Decision map[string]any `json:"decision"`
		Trace    []TraceEntry   `json:"trace"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&out); err != nil {
		return nil, err
	}
	return &Decision{Pack: packID + "@" + version, Attrs: out.Decision, Trace: out.Trace}, nil
}

// Evaluator tries the rules-engine first, then RP_PACKS_DIR, then embedded.
type Evaluator struct {
	Engine   *EngineClient
	PacksDir string // override dir containing packs/<id>/<version>.yaml
}

func (e *Evaluator) Evaluate(packID, version string, ctx map[string]any) (*Decision, error) {
	if e.Engine != nil && e.Engine.BaseURL != "" {
		if d, err := e.Engine.Evaluate(packID, version, ctx); err == nil {
			return d, nil
		}
	}
	p, err := e.loadLocal(packID, version)
	if err != nil {
		return nil, err
	}
	return Evaluate(p, ctx), nil
}

// Load returns the pack document (dir → embedded fallback).
func (e *Evaluator) loadLocal(packID, version string) (*Pack, error) {
	if e.PacksDir != "" {
		if p, err := Load(e.PacksDir, packID, version); err == nil {
			return p, nil
		}
	}
	return LoadEmbedded(packID, version)
}

// LoadPack returns the pack document via the fallback chain.
func (e *Evaluator) LoadPack(packID, version string) (*Pack, error) {
	return e.loadLocal(packID, version)
}

File: packages/py/meridian_py/rulepack.py
"""rp-* rule-pack loading and evaluation (SPEC 1.4), mirroring the Go
packages/shared/rulepack semantics. Fallback chain:
rules-engine/rp-registry API (env) -> RP_PACKS_DIR -> repo embedded packs.
"""

from __future__ import annotations

import os
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any

import httpx
import yaml

_OPS = ("__exists", "__lte", "__gte", "__in", "__ne", "__lt", "__gt")


@dataclass
class Pack:
    id: str
    version: str
    effective_from: str = ""
    effective_to: str | None = None
    status: str = "published"
    subject_to_regazette: bool = True
    provenance: dict = field(default_factory=dict)
    signed: dict = field(default_factory=dict)
    rules: list[dict] = field(default_factory=list)
    raw: dict = field(default_factory=dict)

    @property
    def ref(self) -> str:
        return f"{self.id}@{self.version}"

    @classmethod
    def from_yaml(cls, text: str) -> "Pack":
        doc = yaml.safe_load(text)
        return cls(
            id=doc["id"], version=str(doc["version"]),
            effective_from=str(doc.get("effective_from", "")),
            effective_to=(None if doc.get("effective_to") in (None, "null")
                          else str(doc.get("effective_to"))),
            status=doc.get("status", "published"),
            subject_to_regazette=bool(doc.get("subject_to_regazette", True)),
            provenance=doc.get("provenance") or {},
            signed=doc.get("signed") or {},
            rules=doc.get("rules") or [],
            raw=doc,
        )


def _repo_packs_dir() -> Path | None:
    """Locate packages/shared/rulepack (contains packs/<id>/<version>.yaml)."""
    here = Path(__file__).resolve()
    for parent in here.parents:
        cand = parent / "packages" / "shared" / "rulepack"
        if (cand / "packs").is_dir():
            return cand
    return None


def _match_cond(key: str, want: Any, ctx: dict) -> tuple[bool, str]:
    base, op = key, "eq"
    for suffix in _OPS:
        if key.endswith(suffix):
            base, op = key[: -len(suffix)], suffix[2:]
            break
    got = ctx.get(base)
    exists = base in ctx
    if op == "exists":
        return (exists == bool(want)), f"{base} exists={exists}"
    if op == "in":
        ok = any(str(item) == str(got) for item in (want or []))
        return ok, "" if ok else f"{base}={got} not in {want}"
    if op == "eq":
        if str(got) == str(want):
            return True, ""
        try:
            if float(got) == float(want):
                return True, ""
        except (TypeError, ValueError):
            pass
        return False, f"{base}={got} != {want}"
    if op == "ne":
        ok, _ = _match_cond(base, want, ctx)
        return (not ok), f"{base} ne {want}"
    try:
        gf, wf = float(got), float(want)
    except (TypeError, ValueError):
        return False, f"{base} not numeric"
    ok = {"lt": gf < wf, "lte": gf <= wf, "gt": gf > wf, "gte": gf >= wf}[op]
    return ok, "" if ok else f"{base}={gf} fails {op} {wf}"


def evaluate(pack: Pack, ctx: dict) -> dict:
    """Evaluate a pack over a context -> {pack, decision, trace}."""
    decision: dict[str, Any] = {}
    trace: list[dict] = []
    for rule in pack.rules:
        when = rule.get("when") or {}
        matched, why = True, ""
        for k, want in when.items():
            ok, why = _match_cond(k, want, ctx)
            if not ok:
                matched = False
                break
        trace.append({"rule_id": rule.get("id"), "matched": matched,
                      "reason": "" if matched else why})
        if matched:
            decision.update(rule.get("then") or {})
    return {"pack": pack.ref, "decision": decision, "trace": trace}


class PackRegistry:
    """Loads packs with the fallback chain: rp-registry HTTP API ->
    RP_PACKS_DIR -> repo-embedded copies."""

    def __init__(self, registry_url: str | None = None,
                 packs_dir: str | None = None) -> None:
        self.registry_url = (registry_url
                             or os.environ.get("RP_REGISTRY_URL", "")).rstrip("/")
        self.packs_dir = packs_dir or os.environ.get("RP_PACKS_DIR")
        self._cache: dict[tuple[str, str], Pack] = {}

    def load(self, pack_id: str, version: str = "") -> Pack:
        key = (pack_id, version)
        if key in self._cache:
            return self._cache[key]
        pack = self._load_uncached(pack_id, version)
        self._cache[key] = pack
        return pack

    def _load_uncached(self, pack_id: str, version: str) -> Pack:
        if self.registry_url:
            try:
                if version:
                    url = f"{self.registry_url}/v1/packs/{pack_id}/{version}"
                else:
                    url = f"{self.registry_url}/v1/packs/{pack_id}/latest"
                resp = httpx.get(url, timeout=3.0)
                if resp.status_code == 200:
                    doc = resp.json()
                    if "yaml" in doc:
                        return Pack.from_yaml(doc["yaml"])
                    return Pack.from_yaml(yaml.safe_dump(doc))
            except Exception:
                pass  # fall through to local copies
        for base in ([Path(self.packs_dir)] if self.packs_dir else []) + \
                    ([_repo_packs_dir()] if _repo_packs_dir() else []):
            if base is None:
                continue
            pack_dir = base / "packs" / pack_id
            if not pack_dir.is_dir():
                continue
            if version:
                cand = pack_dir / f"{version}.yaml"
                if cand.is_file():
                    return Pack.from_yaml(cand.read_text())
            else:
                versions = sorted(pack_dir.glob("*.yaml"))
                if versions:
                    return Pack.from_yaml(versions[-1].read_text())
        raise FileNotFoundError(f"rule pack {pack_id}@{version or 'latest'} not found")

    def evaluate(self, pack_id: str, ctx: dict, version: str = "") -> dict:
        """Try core rules-engine (RULES_ENGINE_URL) then local evaluation."""
        engine = os.environ.get("RULES_ENGINE_URL", "").rstrip("/")
        if engine:
            try:
                resp = httpx.post(f"{engine}/v1/evaluate", timeout=3.0, json={
                    "pack_id": pack_id, "version": version, "context": ctx})
                if resp.status_code == 200:
                    body = resp.json()
                    return {"pack": f"{pack_id}@{version}",
                            "decision": body.get("decision", {}),
                            "trace": body.get("trace", []),
                            "via": "rules-engine"}
            except Exception:
                pass
        out = evaluate(self.load(pack_id, version), ctx)
        out["via"] = "embedded-pack"
        return out
